### ISSUE TITLE
FilterAllocationDecider routing exclude setting behavior

### ISSUE THREAD

**Issue by sohami** [elastic/elasticsearch#55764]
**Elasticsearch version** (Verified on 6.8 and 7.4.2):

**Plugins installed**: [none]

**Description of the problem including expected versus actual behavior**:
I was exploring the FilterAllocationDecider behavior with **cluster.routing.allocation.exclude** setting, specially how the updates of these setting group works. I found that the behavior is not as per my expectation and think it is a bug in ES.

Let's say there are 2 attributes "attr_1" and "attr_2" for which we want to put the exclusion filter for FilterAllocationDecider to consider. FilterAllocationDecider has a consumer registered for the settings in **cluster.routing.allocation.exclude** group. The expectation is with each update setting request for this group, the clusterExcludeFilters in FilterAllocationDecider will be updated such that keeps all the exclusion applied so far. However, it looks like it only keeps the difference in previous and current settings for this group.

**Steps to reproduce**:
Case 1:
- Request 1: set the value of only "attr_1" to be "1"
   - Verified `clusterExcludeFilter` in `FilterAllocationDecider` has key: "attr_1" with value: "1"
- Request 2: set the value of only "attr_2" to be "abc"
   - Expectation is `clusterExcludeFilter` in `FilterAllocationDecider` will have 2 keys "attr_1" and "attr_2" with value "1" and "abc" respectively. However, the second request removes the first key "attr_1" and only keep the second key "attr_2" in the `clusterExcludeFilter`

Case 2:
- Request 1: set the value of "attr_1" to be "1" and "attr_2" to be "abc"
   - Verified `clusterExcludeFilter` in `FilterAllocationDecider` has key: "attr_1" with value: "1" and key "attr_2" with value: "abc"
- Request 2: set the value of "attr_1" to be "1" and "attr_2" to be "def"
   - Expectation is `clusterExcludeFilter` in `FilterAllocationDecider` will have 2 keys "attr_1" and "attr_2" with value "1" and "def" respectively. However, the second request removes the first key "attr_1" and only keep the second key "attr_2" with value "def" in the `clusterExcludeFilter`

I have created a test here for various scenarios showing the behavior: 
https://github.com/sohami/elasticsearch/commits/7.4.2-test

**Details:**
For **cluster.routing.allocation.exclude** setting the FilterAllocationDecider registers a setting updater created using [newAffixMapUpdater](https://github.com/sohami/elasticsearch/blob/7.4.2-test/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L736). This updater checks for the difference in settings for this group between current and previous settings and calls the consumer with the diff of the setting. There is a local [result](https://github.com/sohami/elasticsearch/blob/7.4.2-test/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L748) map which is populated with the changed setting and passed to the [consumer here](https://github.com/sohami/elasticsearch/blob/7.4.2-test/server/src/main/java/org/elasticsearch/common/settings/Setting.java#L766). 

**Proposal:**
- The result map in the `newAffixMapUpdater` should be a member variable rather than local variable, so that it keeps all the changed settings so far and always return that to the consumer.
- `FilterAllocationDecider` keeps adding/updating the settings as and when updated instead of reinitializing the new [DiscoveryNodeFilters](https://github.com/sohami/elasticsearch/blob/7.4.2-test/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java#L196) for each setting update

**Comment by DaveCTurner** [comment]
I think this is more of a user question than a genuine bug report and as such I am closing it here. Please ask this question on the [discussion forums](https://discuss.elastic.co/c/elasticsearch) instead.

**Comment by sohami** [comment]
@DaveCTurner - I have opened a post in [discussion forum here](https://discuss.elastic.co/t/filterallocationdecider-routing-exclude-setting-behavior/229800). However, I was actually pointing out to a bug in the exclude setting behavior which is not as expected. I have also shared a test to show the existing behavior and what is the expectation. 

I am new to the community so just wanted to know if the process is that based on discussion in the forum it will be decided if this is a bug or not ?
